File: flink-utils/src/main/scala/io/github/ppdzm/utils/flink/serde/CachedSchemaCoderProvider.scala
package io.github.ppdzm.utils.flink.serde

import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient
import io.github.ppdzm.utils.universal.implicits.BasicConversions._
import org.apache.flink.formats.avro.SchemaCoder
import org.apache.flink.formats.avro.registry.confluent.ConfluentSchemaRegistryCoder

@SerialVersionUID(4023134423033312666L)
class CachedSchemaCoderProvider(subject: String, url: String, identityMapCapacity: Int) extends SchemaCoder.SchemaCoderProvider {

    override def get: SchemaCoder = {
        if (subject.notNullAndEmpty)
            new ConfluentSchemaRegistryCoder(subject, new CachedSchemaRegistryClient(url, identityMapCapacity))
        else
            new ConfluentSchemaRegistryCoder(new CachedSchemaRegistryClient(url, identityMapCapacity))
    }

}

File: flink-utils/src/main/java/io/github/ppdzm/utils/flink/common/CachedSchemaCoderProvider.java
package io.github.ppdzm.utils.flink.common;

import io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient;
import org.apache.flink.formats.avro.SchemaCoder;
import org.apache.flink.formats.avro.registry.confluent.ConfluentSchemaRegistryCoder;

/**
 * @author Created by Stuart Alex on 2021/5/7.
 */
public class CachedSchemaCoderProvider implements SchemaCoder.SchemaCoderProvider {
    private static final long serialVersionUID = 4396760227837662376L;
    private String subject;
    private String schemaRegistryUrl;
    private int identityMapCapacity;

    public CachedSchemaCoderProvider(String subject, String schemaRegistryUrl, int identityMapCapacity) {
        super();
        this.subject = subject;
        this.schemaRegistryUrl = schemaRegistryUrl;
        this.identityMapCapacity = identityMapCapacity;
    }

    @Override
    public SchemaCoder get() {
        if (subject == null || subject.isEmpty()) {
            return new ConfluentSchemaRegistryCoder(subject, new CachedSchemaRegistryClient(schemaRegistryUrl, identityMapCapacity));
        } else {
            return new ConfluentSchemaRegistryCoder(new CachedSchemaRegistryClient(schemaRegistryUrl, identityMapCapacity));
        }
    }
}
